---
title: "Next-generation sequencing - analysis with DESeq2"
subtitle: "Comparison of gene expression in DSCs compared to melanocytes"
author: "Marc Bender"
date: '`r Sys.Date()`'
output: 
  word_document:
    toc: true
  pdf_document:
    toc: true
    latex_enginge: xelatex
  html_notebook:
    theme: united
    toc: true
  html_document:
    df_print: paged
    toc: true
    code_folding: "hide"
header-includes: \usepackage[utf8]{inputenc}
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  echo    = FALSE
)
```

```{r load_libraries, include = FALSE}
library(DESeq2)
library(tidyverse)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(apeglm)
library(EnhancedVolcano)
library(pheatmap)
library(ComplexHeatmap)
library(ReportingTools)
library(PoiClaClu)
library(msigdbr)
library(clusterProfiler)
library(flextable)
library(GGally)
library(ggprism)
```

```{r custom_functions, include = FALSE}
## function to print input thresholds
print_input <- function(SummarizedExp, DESeqObj, pThres = 0.05, lfcThres = 1, minimumCount = 10, 
                        smallesGroupSize, contrast){
  ft <- data.frame(Thresholds = NA, pThres, lfcThres, minimumCount, smallestGroupSize, formula = paste(as.character(contrast), collapse = ""),
                   Transcripts = NA, n_total = nrow(assay(se)), n_expr = nrow(assay(DESeqObj))) %>%
    gather(rownames, value) %>%
    mutate(rownames = factor(rownames, levels = c("Thresholds", "pThres", "lfcThres", "minimumCount",
                                                  "smallestGroupSize",  "formula", "Transcripts", "n_total", "n_expr"), labels = c("Thresholds", "P-value", "Log-fold change", "Minimum read count", "Smallest group size",                             "Design Formula", "Transcript number",  "Total number of transcripts",
                                                                                                                                   "Filtered number of transcripts"))) %>%
    flextable() %>%
    set_header_labels(rownames = "", value = "Value") %>%
    bold(part = "header") %>%
    bold(i = c(1,6,7), j = 1) %>%
    padding(i = c(2,3,4,5,8,9), j = 1, padding.left = 10)%>% 
    line_spacing(space = 0.4, part = "body") %>%
    footnote(i = 9, j = 1,
             value = as_paragraph(
               c(paste("All transcripts which are above the minimum read count threshold (", minimumCount,") in at 
                      least ", smallestGroupSize, " samples.", sep = ""))
             ),
             ref_symbols = c("\u2020"),
             part = "body") %>%
    fontsize(size = 9, part = "all") 
  
  emR::FitFlextableToPage(ft)
}

## function to print reads per sample
print_reads <- function(readsRaw, readsFiltered){
  nreads <- as.data.frame(rbind(formatC(nread, format = "e", digits = 3), 
                                formatC(nreadFiltered, format = "e", digits = 3)))
  rownames <- c("Number of reads (total)", "Number of reads (filtered)")
  
  # transform into ngsekb function
  ft <- cbind(rownames, nreads) %>% flextable() %>%
    set_header_labels(rownames = "") %>%
    bold(part = "header") %>%
    footnote(i = 2, j = 1,
             value = as_paragraph(
               c(paste("All transcripts which are above the minimum read count threshold (", minimumCount,") in at 
                      least", smallestGroupSize, " samples.", sep = ""))
             ),
             ref_symbols = c("\u2020"),
             part = "body") %>%
    fontsize(size = 9, part = "all") 
  
  emR::FitFlextableToPage(ft)
}

## function to plot PCA plots
pca_plot <- function(type){
  ggplot(pcaData, aes_string(x = "PC1", y = "PC2", color = factor1, shape = factor2)) +
    geom_point(size =3) +
    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    ggtitle(paste("PCA with", type,"data")) +
    theme_bw() + 
    ggsci::scale_color_npg() 
}

## function to specify lower plot panel in ggpairs function
lowerFn <- function(data, mapping,...){
  p <- ggplot(data = data, mapping = mapping) + 
    geom_point() +
    geom_abline(intercept = 0, slope = 1, color = "red") +
    scale_x_continuous(breaks = seq(floor(min(data)/2)*2, ceiling(max(data)/2)*2, 2)) +
    scale_y_continuous(breaks = seq(floor(min(data)/2)*2, ceiling(max(data)/2)*2, 2))
}
```

```{r load_data}
setwd("Z:/Aktuell/Eigene Dateien/Eigene Dateien_Marc/R/Github/EKB/DSC_NGS/")
# source(here::here("R/DSC_NGS.R"))
load(file = "DSC_NGS.RData")
warning("Please check carefully if correct RData object has been loaded.")
```

\newpage 
# Resources

For additional info on analysis workflows check: 
https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html 
https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html  
https://www.bigbioinformatics.org/r-and-rnaseq-analysis 
https://github.com/bigbioinformatics/r-programming-and-rnaseq-workshop 

\newpage
# Thresholds, design formula and number of transcripts

```{r print_input}
# transform into ngsekb function
print_input(SummarizedExp = se, DESeqObj = dds, pThres = pThres, lfcThres = lfcThres, minimumCount = minimumCount, 
            smallesGroupSize = smallestGroupSize, contrast = contrast)
```

\newline 

```{r print_reads}
print_reads(nread, nreadFiltered)
```

\newpage
# Sample sheet data

\newpage
# Exploratory data analysis

## Heatmap of Poisson distances between samples

* IMPORTANT: 
  + use Poisson distance for raw (non-normalized) count data
  + use Euclidean distance for data normalized by regularized-logarithm transformation (rlog) or variance stablization transfromation (vst)

```{r poisd}
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colorsPoisd)
poisd <- NULL # remove big poisd object
se <- NULL
```


\newpage
## Transform data 

Which transformation to choose? The VST is much faster to compute and is less sensitive to high count outliers
than the rlog. The rlog tends to work well on small datasets (n < 30), potentially outperforming the VST 
when there is a wide range of sequencing depth across samples (an order of magnitude difference). 
We therefore recommend the VST for medium-to-large datasets (n > 30). You can perform both transformations and 
compare the meanSdPlot or PCA plots generated, as described below.

To show the effect of the transformation, in the figure below we plot the first sample against the second,
first simply using the log2 function (after adding 1, to avoid taking the log of zero), and then using the 
VST and rlog-transformed values. For the log2 approach, we need to first estimate size factors to account for
sequencing depth, and then specify normalized=TRUE. Sequencing depth correction is done automatically for the 
vst and rlog.

```{r plot_transformations}
# build data set
df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
    mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))

# assign column names
colnames(df)[1:2] <- c("x", "y")

# transform to factor
lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)

# plot values
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation) 
```

\newpage
## Principal component analysis (PCA)

Die Hauptkomponentenanalyse oder *principal component analysis* (PCA) ist ein Verfahren zur Reduktion der Datendimensionalität und findet Anwendung bei komplexen Datensätzen (aus n Proben mit p gemessenen Merkmalen), um Muster oder Strukturen in den Daten zu erkennen und die Interpretation der Daten zu vereinfachen. Dies erfolgt durch Identifikation neuer unkorrelierter Variablen, den sog. Hauptkomponenten (PC = *principal component*), welche sukzessive die erklärte Varianz innerhalb eines Datensets maximieren. Die Berechnung dieser Hauptkomponenten erfolgt dabei durch Zentrieren der Daten und einen nachfolgenden iterativen Prozess, bei dem eine zufällige Gerade durch den Mittelwert der Datenpunkte gelegt und so lange rotiert wird, bis der Abstand der auf die Gerade projizierten Datenpunkte zum Zentrum maximal ist. 
Die Distanzen werden berechnet als:

$$
\begin{align*}
SS(distances) = d^2_{1} + d^2_{2} + ... + d^2_{n}
\end{align*}
$$
					 
				
mit SS(distances) = Eigenwert (Quadratsumme der Distanzen), d = Distanz vom projizierten Datenpunkt zum Zentrum, n = Probengröße. 
					
Die Gerade mit dem größten Eigenwert wird als PC1 bezeichnet und trägt am meisten zur Varianzaufklärung bei. Im Anschluss wird die nächste Gerade gesucht, die orthogonal zur ersten Gerade ist und ebenfalls durch den Mittelwert der Datenpunkte verläuft. Nach diesem Grundsatz können (bis zur n-ten Hauptkomponente) weitere Hauptkomponenten gefunden werden, wobei diese stets orthogonal zu den bisherigen Hauptkomponenten stehen müssen. 
					
Die Reduktion der Dimensionalität beruht darauf, dass die jeweiligen Hauptkomponenten einen unterschiedlich großen (absteigend von PC1 zu PCn) Anteil der Datenvarianz erklären und es daher in den meisten Fällen ausreicht die ersten drei Hauptkomponenten zu analysieren (welche jeweils als 2-dimensionaler Plot gegeneinander aufgetragen werden), um Gemeinsamkeiten innerhalb der Proben zu identifizieren. 			
				Der Anteil der Varianz, der durch die jeweilige Hauptkomponente erklärt wird ($var_{explained}$), kann anhand folgender Formel  berechnet werden:					
				
				
$$
\begin{align*}
var(PC_{i}) = \frac{SS(distances (PC_{i}))}{n-1}
\end{align*}
$$ 
				
				
$$
\begin{align*}
var_{explained} = \frac{var(PC_{i})}{\sum^{n}_{j=1}var(PC_{j})}
\end{align*}
$$
				
				
mit $PC_{i}$ = Hauptkomponente i. \newline
					
Die PCA wurde in dem Statistikprogramm R mithilfe der Pakete *factoextra* und *FactoMineR* durchgeführt. Die Ermittlung relevanter Hauptkomponenten erfolgte mittels Skree Plot. Dort werden entweder die Eigenwerte oder die erklärte Varianz gegen die Nummer der jeweiligen Hauptkomponente aufgetragen und mittels einer zunächst steil abfallenden Linie, die sich asymptotisch der Abzisse annähert, verbunden. An dem sog. *elbow* ist ein deutlicher Knick in der Linie zu erkennen. Faktoren, die links dieses Punktes liegen reichen größtenteils aus, um Muster in den Daten zu erklären. Die anderen Faktoren unterscheiden sich oft nicht deutlich von Zufallskorrelationen und sind daher nicht von Bedeutung. 

```{r PCA, dpi = 300, fig.width=8, fig.height= 3}
## plot pca for vst data

## plot PCA
pcaData <- plotPCA(vsd, intgroup = c(factor1, factor2), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

## plot data with ggplot
pcaVST <- pca_plot(type = "VST")

## plot pca for rlog data
pcaData <- plotPCA(rld, intgroup = c(factor1, factor2), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# plot data with ggplot
pcaRlog <-  pca_plot(type = "Rlog")

## arrange plots
ggpubr::ggarrange(pcaVST, pcaRlog, nrow = 1, common.legend = TRUE) 
```

\newpage
## Scatter plot matrices 

These types of plots were suggested in https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2968-1.
They depict the correlation pairs of each sample with other samples. The spread of points should be closer to the identity line (red line) in technical replicates than in biological replicates because a larger spread indicates higher differences in expression values most often associated with differential expression. Divergence of the expected pattern (as described in the aforemented publication) should also warrant a closer inspection. 

```{r scatter_matrix, dpi = 300, fig.width=22, fig.height= 22}
ggpairs(data = as.data.frame(assay(rld)), 
        lower = list(continuous = wrap(lowerFn)),
        upper = list(continuous = wrap(ggally_cor, stars = F)),
        xlab = c("Rlog counts"), ylab = "Rlog counts") +
        theme_classic(base_size = 26)
```

\newpage
# Differential expression analysis

## Volcano Plot (most differentially expressed genes labelled)
<!-- Text explaining Volcano Plots and interpretation. -->

This plot shows the most significantly changed genes (<= 1e-30, logFC >= |8|). 

```{r Volcano_plot_sig, dpi = 300, fig.width=18, fig.height= 12}
# 5. Get the results and plot Volcano plot without filter within results function
res_volcano <- results(dds)

## annotate genes
anno <- AnnotationDbi::select(org.Hs.eg.db, rownames(res_volcano),
                              columns=c("ENSEMBL", "ENTREZID", "SYMBOL", "GENENAME"),
                              keytype="ENSEMBL")

# adding ENSEMBL gene ID as a column in significant differentially expression gene table.
dat_volcano <- cbind(ENSEMBL = rownames(res_volcano), res_volcano)
dat_volcano <- left_join(as.data.frame(dat_volcano), anno)
res_volcano <- NULL

# plot Volcano
# this only serves as diagnostic and representation tool. the actual fold change thresholds and pvalues change later due to the way the results function works
EnhancedVolcano(as.data.frame(dat_volcano), lab = dat_volcano$SYMBOL, selectLab = ifelse(dat_volcano$padj <= 1e-30 & abs(dat_volcano$log2FoldChange) >=8, dat_volcano$SYMBOL, NA),
                x = 'log2FoldChange', y = 'padj', gridlines.major = FALSE, gridlines.minor = FALSE,
                xlim = c(-21, 21), ylim = c(-5, 200), title = 'Melanocytes vs. DSCs', 
                pCutoff = pThres, FCcutoff = lfcThres, pointSize = 3, labSize = 5) 
```

## Volcano Plot (skin cell marker genes labelled)

This plot shows 24 selected genes from a publication assessing differences between Melanocytes (12 genes) and DSCs (12 genes) (Zabierowski et al. 2012) as well as 12 genes characteristic for Fibroblasts. 

```{r Volcano_plot_markers, dpi = 300, fig.width=18, fig.height= 12}
## define marker genes based on publication by Zabierowski 2011
mcMarkers <- c("MLANA", "PMEL", "TYR", "TYRP1", "MITF", "DCT",  "KIT", "CDH1", "ERBB3", "EDNRB", "SOX10", "PAX3")
dscMarkers <-  c("SNAI1", "NGFR", "NES", "FRZB", "FOXC1", "JAG1", "PAX3", "HES4", "MSX1", "S100A4", "SOX2", "CDH2")
fbMarkers <- c("THY1", "DLK1", "FAP", "PDGFRA", "ELN", "FN1", "VIM", "LAMA5", "ITGB1", "DPP4", "SERPINH1", "LAMA2")

markerGenes <- c(mcMarkers, dscMarkers, fbMarkers)

## Change colors based on which marker and remove the names (maybe)

EnhancedVolcano(as.data.frame(dat_volcano), lab = dat_volcano$SYMBOL, selectLab = markerGenes,
                x = 'log2FoldChange', y = 'padj', gridlines.major = FALSE, gridlines.minor = FALSE,
                xlim = c(-21, 21), ylim = c(-5, 200), title = 'Melanocytes vs. DSCs',
                pCutoff = pThres, FCcutoff = lfcThres, pointSize = 3, labSize = 5)
```

## Volcano Plot (melanoma marker genes labelled)

This plot shows genes frequently mutated in melanoma (taken from Cherepakhin et. al (2022)). 

```{r Volcano_plot_melanoma, dpi = 300, fig.width=18, fig.height= 12}
## define marker genes based on publication by Cherepakhin et. al (2022)
melMarkers <- c("BRAF", "NRAS", "MAP2K1", "CTNNB1", "APC", "BAP1", "PRKAR1A", "PRKCA", "TERT", "CDKN2A", "TP53", "PTEN", "RAC1", "NF1", "KIT", "NFKBIE", "PIK3CA", "PTPN11", "ERBB2", "MAP3K1", "EGFR", "MET", "HRAS", "ALK", "RET", "ROS1", "NTRK1", "NTRK3", "CCND1", "GAB2")

# change into colored points 

EnhancedVolcano(as.data.frame(dat_volcano), lab = dat_volcano$SYMBOL, selectLab = melMarkers,
                x = 'log2FoldChange', y = 'padj', gridlines.major = FALSE, gridlines.minor = FALSE,
                xlim = c(-21, 21), ylim = c(-5, 200), title = 'Melanocytes vs. DSCs',
                pCutoff = pThres, FCcutoff = lfcThres, pointSize = 3, labSize = 5)
```

\newpage
## MA plot

An MA-plot (Dudoit et al. 2002) provides a useful overview for the distribution of the estimated
coefficients in the model, e.g. the comparisons of interest, across all genes. On the y-axis, the “M”
stands for “minus” – subtraction of log values is equivalent to the log of the ratio – and on the x-axis,
the “A” stands for “average”. You may hear this plot also referred to as a mean-difference plot, or a
Bland-Altman plot.

Before making the MA-plot, we use the lfcShrink function to shrink the log2 fold changes for the comparison
of DSCs vs. Melanocytes. There are three types of shrinkage estimators in DESeq2, which are covered
in the DESeq2 vignette. Here we specify the apeglm method for shrinking coefficients, which is good for shrinking
the noisy LFC estimates while giving low bias LFC estimates for true large differences (Zhu, Ibrahim, and Love 2018).
To use apeglm we specify a coefficient from the model to shrink, either by name or number as the coefficient appears
in resultsNames(dds).

WARNING: DO NOT ALWAYS SHRINK EVERYTHING BY DEFAULT. Look at MA plot and Volcano plot to see if it makes sense.

```{r MA_plot, dpi = 300, fig.width=5, fig.height= 4}
### plot MA plot

# plot with original lfc
plotMA(res, ylim = c(-5, 5), main = "MA plot with original log fold changes")

# shrink estimates and exlpore differences in MA plots
plotMA(resShrunk, ylim = c(-5, 5), main = "MA plot with shrunken log fold changes")

###
# WARNING: DO NOT ALWAS SHRINK EVERYTHING BY DEFAULT. Look at MA plot and Volcano plot to see if it makes sense
### 
```

\newpage
## Comparison of marker genes

## Expression of melanocyte markers

Marker genes were labelled according to the publication of Zabierowski 2012, lab practice and https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4278762/. 

```{r mc_marker_genes, dpi = 300, fig.width=7, fig.height= 8}
marker_gene_plot <- function(gene.ids){
  markerInd   <- which(allRes$SYMBOL %in% gene.ids)
  markerRld <- assay(rld)[markerInd,]
  markerDEG <- allRes[markerInd,]
  rownames(markerRld) <- markerDEG$SYMBOL
  
  markerPlotDf <- markerRld %>%
  as.data.frame() %>% 
  rownames_to_column("SYMBOL") %>%
  gather("ID", "RLOG", -"SYMBOL") %>%
  mutate(cell_type = ifelse(str_detect(ID, "DSC"), "DSC", "Melanocytes")) %>%
  left_join(markerDEG[,c("SYMBOL", "padj")]) %>%
  mutate(padj = paste("p.adj =",formatC(padj, format = "e", digits = 2)))
  markerPlotDf[-c(1:length(gene.ids)),]$padj <- NA
  
  markerPlotDf %>%
  ggplot(aes(x= cell_type, y = RLOG, fill = cell_type)) +
  geom_bar(stat = "summary", fun = "mean", color = "black") +
  geom_text(aes(label = padj), x =1.5, y = 21, size = 3)+
  geom_point(size = 2, position = position_jitterdodge()) +
  facet_wrap(~SYMBOL, scales = "free_y") +
  ggsci::scale_fill_npg() +
  scale_x_discrete(guide = "prism_offset") +
  scale_y_continuous(limits = c(0,21), guide = "prism_offset") +
  xlab("") +
  ylab("Rlog counts") +
  guides(fill=guide_legend(title="Cell type")) +
  theme_prism(base_size = 12) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
}

marker_gene_plot(mcMarkers)
```

\newpage
## Expression of dsc markers

```{r dsc_marker_genes, dpi = 300, fig.width=7, fig.height= 8}
marker_gene_plot(dscMarkers)
```

\newpage
## Expression of fibroblast markers

```{r fb_marker_genes, dpi = 300, fig.width=7, fig.height= 8}
marker_gene_plot(fbMarkers)
```

\newpage
## Expression of genes associated with melanomgenesis

The following genes were extracted from a publication of Cherepakhin et. al (2022) (https://doi.org/10.3390/cancers14010123) for cutaneous melanoma entities. 

```{r melanoma_markers_png, echo=FALSE, out.width = '100%'}
knitr::include_graphics("/Results/Melanoma_Mutation_Markers.png", error = FALSE)
```

\newpage
```{r melanoma_marker_genes, dpi = 300, fig.width=10, fig.height= 13}
marker_gene_plot(melMarkers)
```

\newpage
## Heatmap

If using either of Euclidean distance or Pearson correlation, data should follow a Gaussian /
normal (parametric) distribution. So, if coming from a microarray, anything from RMA normalisation is fine,
whereas, if coming from RNA-seq, any data deriving from a transformed normalised count metric should be fine,
such as variance-stabilised, regularised log, or log CPM expression levels.

If you are performing clustering on non-normal data, like 'normalised' [non-transformed] RNA-seq counts,
FPKM expression units, etc., then use Spearman correlation (non-parametric).

\newpage
### All differentially expressed (DE) genes
```{r Heatmap, dpi = 300, fig.width=12, fig.height= 12}
# Heatmap of all DE genes
matAll <- assay(rld)[sigRes$ENSEMBL,]

## specify annotation bar
annoHeatmap <- as.data.frame(colData(rld)[, c("cell_type")])
colnames(annoHeatmap) <- "Cell_type"

## specify annotation bar colors
annColors <- list(Cell_type = c(DSC = ggsci::pal_npg("nrc")(2)[1], Melanocytes = ggsci::pal_npg("nrc")(2)[2]))

## print heatmap
ComplexHeatmap::pheatmap(matAll, scale = "row", show_rownames = FALSE, clustering_distance_rows = "correlation",
         annotation_colors = annColors, heatmap_legend_param = list(title = "Row z-score", at = c(-2,1,0,1,2)),
         annotation_col = annoHeatmap,  main= paste("Differentially Expressed genes (", nrow(sigRes), ")", sep =""))
```

\newpage
### Top 30 DE genes
```{r HeatmapTop30, dpi = 300, fig.width=12, fig.height= 11}
nGenes <- 30
topSignif <- head(order(allRes$padj, decreasing = FALSE),nGenes)
labelsTop30 <- allRes[topSignif,]$SYMBOL
matTop30  <- assay(rld)[topSignif, ]

ComplexHeatmap::pheatmap(matTop30,  annotation_col = annoHeatmap, labels_row = labelsTop30, 
                         clustering_distance_rows = "correlation",  annotation_colors = annColors, 
                         heatmap_legend_param = list(title = "Rlog counts", at = c(5,10,15,20)),
                         main= paste("Top ",nGenes," differentially expressed genes", sep =""))
```


\newpage
### Comparison of genes identified with the DRAGEN pipeline

This plot includes the same 30 genes which are depicted in the Heatmap included in the output generated from the DRAGEN analysis software. 

```{r HeatmapDragen, dpi = 300, fig.width=12, fig.height= 11}
ind <- which(allRes$SYMBOL %in% DragenRes$gene_name)
labelsDragen <- allRes[ind,]$SYMBOL
matDragen  <- assay(rld)[ind, ]

ComplexHeatmap::pheatmap(matDragen,  annotation_col = annoHeatmap, labels_row = labelsDragen, 
                         clustering_distance_rows = "correlation",  annotation_colors = annColors, 
                         heatmap_legend_param = list(title = "Rlog counts", at = c(5,10,15,20)),
                         main= paste("DE genes identified with the DRAGEN pipeline", sep =""))
```


\newpage
# Functional analysis

## Overrepresentation analysis (mirsigdb hallmark pathways)

Gene sets are extracted from https://www.gsea-msigdb.org/gsea/msigdb/ (R package msigdbr). Enrichment is done with the R package ClusterProfiler (If you use clusterProfiler in published research, please cite:
T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141)).

Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying overlaps between gene sets in other MSigDB collections and retaining genes that display coordinate expression.

```{r define_Hallmark_GS}
overExpressed <- sigRes %>%
  filter(log2FoldChange > 0) %>%
  pull(SYMBOL)

underExpressed <- sigRes %>%
  filter(log2FoldChange < 0) %>%
  pull(SYMBOL)

geneSets <- msigdbr(species = "Homo sapiens", category = "H")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)
```

### Hallmark: upregulated genes
```{r hallmark_up, dpi = 300, fig.width=13, fig.height= 12}
## Enrichment GO biological process
upReg_Hallmark <- enricher(gene = overExpressed, TERM2GENE = geneSets)
dotplot(upReg_Hallmark, showCategory = 20)
```

\newpage
### Hallmark: downregulated genes
```{r hallmark_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
downReg_Hallmark <- enricher(gene = underExpressed, TERM2GENE = geneSets)
dotplot(downReg_Hallmark, showCategory = 20)
```

\newpage
## Overrepresentation analysis GO Terms

Gene sets that contain genes annotated by the same ontology term. The C5 collection is divided into two subcollections, the first derived from the Gene Ontology resource (GO) which contains BP, CC, and MF components and a second derived from the Human Phenotype Ontology (HPO).

### Analysis of overexpressed genes
```{r define_upreg}
geneSets <- msigdbr(species = "Homo sapiens", category = "C5")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)
```

\newpage
#### GO: biological process
```{r GOBP_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
geneSets_GOBP <- geneSets[str_detect(geneSets$gs_name, "GOBP"),]
upReg_GOBP <- enricher(gene = overExpressed, TERM2GENE = geneSets_GOBP)
dotplot(upReg_GOBP, showCategory = 20)
```

\newpage
#### GO: molecular function
```{r GOMF_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO molecular function
geneSets_GOMF <- geneSets[str_detect(geneSets$gs_name, "GOMF"),]
upReg_GOMF <- enricher(gene = overExpressed, TERM2GENE = geneSets_GOMF)
dotplot(upReg_GOMF, showCategory = 20)
```

\newpage
#### GO: cellular compartment
```{r GOCC_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO cellular compartments
geneSets_GOCC <- geneSets[str_detect(geneSets$gs_name, "GOCC"),]
upReg_GOCC <- enricher(gene = overExpressed, TERM2GENE = geneSets_GOCC)
dotplot(upReg_GOCC, showCategory = 20)
```

\newpage
### Analysis of underexpressed genes

#### GO: biological process
```{r GOBP_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
geneSets_GOBP <- geneSets[str_detect(geneSets$gs_name, "GOBP"),]
downReg_GOBP <- enricher(gene = underExpressed, TERM2GENE = geneSets_GOBP)
dotplot(downReg_GOBP, showCategory = 20)
```

\newpage
#### GO: molecular function
```{r GOMF_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO molecular function
geneSets_GOMF <- geneSets[str_detect(geneSets$gs_name, "GOMF"),]
downReg_GOMF <- enricher(gene = underExpressed, TERM2GENE = geneSets_GOMF)
dotplot(downReg_GOMF, showCategory = 20)
```

\newpage
#### GO: cellular compartment
```{r GOCC_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO cellular compartments
geneSets_GOCC <- geneSets[str_detect(geneSets$gs_name, "GOCC"),]
downReg_GOCC <- enricher(gene = underExpressed, TERM2GENE = geneSets_GOCC)
dotplot(downReg_GOCC, showCategory = 20)
```

\newpage
## Overrepresentation analysis Regulatory elements (transcription factor binding and miRNAs)

### Regulatory: upregulated genes
```{r reg_up, dpi = 300, fig.width=13, fig.height= 11}
geneSets <- msigdbr(species = "Homo sapiens", category = "C3")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)

## Enrichment GO biological process
upReg_reg <- enricher(gene = overExpressed, TERM2GENE = geneSets)
dotplot(upReg_reg, showCategory = 20)
```

\newpage
### Regulatory: downregulated genes
```{r reg_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
downReg_reg <- enricher(gene = underExpressed, TERM2GENE = geneSets)
dotplot(downReg_reg, showCategory = 20)
```

\newpage
## Gene set enrichment analysis (example)

This analysis is just an example to show the possibility of GSEA and depcition of GSEA plots if preferred to overrepresentation analysis.

```{r gsea}
## Deal with inf
allResGSEA <- allRes %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange))

## Remove NAs and order by GSEA
allResGSEA <- allResGSEA %>%
  filter(! is.na(gsea_metric)) %>%
  arrange(desc(gsea_metric))

## Get the ranked GSEA vector
ranks <- allResGSEA %>%
  select(SYMBOL, gsea_metric) %>%
  distinct(SYMBOL, .keep_all = TRUE) %>%
  deframe()

## define Gene sets (GO)
geneSets <- msigdbr(species = "Homo sapiens", category = "C5")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)

## Run GSEA
gseaRes <- GSEA(geneList = ranks, TERM2GENE = geneSets)
gseaResDf <- as.data.frame(gseaRes)
```

<!-- #### GO: biological process -->
<!-- ```{r GOBP_down, dpi = 300, fig.width=13, fig.height= 11} -->
<!-- ## Enrichment GO biological process -->
<!-- geneSets_GOBP <- geneSets[str_detect(geneSets$gs_name, "GOBP"),] -->
<!-- downReg_GOBP <- enricher(gene = underExpressed, TERM2GENE = geneSets_GOBP) -->
<!-- dotplot(downReg_GOBP, showCategory = 20) -->
<!-- ``` -->

```{r gsea_dotplot, dpi = 300, fig.width=10, fig.height= 12}
## Make summary plots
dotplot(gseaRes, showCategory = 20)
```

```{r gsea_testplot, dpi = 300, fig.width=7, fig.height= 5}
gseaplot(gseaRes, geneSetID = "GOCC_PIGMENT_GRANULE",
         title = "GOCC_PIGMENT_GRANULE")
```

\newpage
# Appenix

## Statistical methods

TBD

## Session Info
```{r}
utils::sessionInfo()
```
