---
title: "Next-generation sequencing - analysis with DESeq2"
subtitle: "Comparison of gene expression in DSCs compared to melanocytes - functional analysis"
author: "Marc Bender"
date: '`r Sys.Date()`'
knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_dir = "../Results",
      output_format = "word_document"
    )
  })
output: 
  word_document:
    toc: true
  pdf_document:
    toc: true
    latex_enginge: xelatex
  html_notebook:
    theme: united
    toc: true
  html_document:
    df_print: paged
    toc: true
    code_folding: "hide"
header-includes: \usepackage[utf8]{inputenc}
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  echo    = FALSE
)
```

```{r load_libraries, include = FALSE}
library(DESeq2)
library(tidyverse)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
```

```{r custom_functions, include = FALSE}
## wrapper to plot GO terms
plot_GO <- function(geneSetRef, geneSet, category, title = "Dotplot", ...){
  gs <- geneSetRef[str_detect(geneSetRef$gs_name, category),]
  enriched <- enricher(gene = geneSet, TERM2GENE = gs)
  dotplot(enriched, label_format = Inf, ...) + ggtitle(title)
}
```

```{r load_data}
setwd("Z:/Aktuell/Eigene Dateien/Eigene Dateien_Marc/R/Github/EKB/DSC_NGS/")
load(file = "DSC_NGS.RData")
```

\newpage 
# Resources and introduction

This document includes the functional analysis for the comparison of RNASeq data from dermal stem cells (DSCs) and Melanocytes. The exploratory data analysis and differential expression analysis are included in separate files. Statistical methods are described in a separate .docx file. 

For additional info on analysis workflows check: 
DESEQ workflow: https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html 
https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html  
https://www.bigbioinformatics.org/r-and-rnaseq-analysis 
https://github.com/bigbioinformatics/r-programming-and-rnaseq-workshop 
ClusterProfiler: https://pubmed.ncbi.nlm.nih.gov/34557778/ 

\newpage
# Overrepresentation analysis

## mirsigdb hallmark pathways

Gene sets are extracted from https://www.gsea-msigdb.org/gsea/msigdb/ (R package msigdbr). Enrichment is done with the R package ClusterProfiler (If you use clusterProfiler in published research, please cite:
T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141)).

Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying overlaps between gene sets in other MSigDB collections and retaining genes that display coordinate expression.

```{r define_Hallmark_GS}
overExpressed <- sigRes %>%
  filter(log2FoldChange > 0) %>%
  pull(SYMBOL)

underExpressed <- sigRes %>%
  filter(log2FoldChange < 0) %>%
  pull(SYMBOL)

geneSets <- msigdbr(species = "Homo sapiens", category = "H")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)
```

### Hallmark: upregulated genes
```{r hallmark_up, dpi = 300, fig.width=13, fig.height= 12}
## Enrichment GO biological process
upReg_Hallmark <- enricher(gene = overExpressed, TERM2GENE = geneSets)
dotplot(upReg_Hallmark, showCategory = 20, label_format = Inf)
```

\newpage
### Hallmark: downregulated genes
```{r hallmark_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
downReg_Hallmark <- enricher(gene = underExpressed, TERM2GENE = geneSets)
dotplot(downReg_Hallmark, showCategory = 20, label_format = Inf)
```

\newpage
## GO Terms

Gene sets that contain genes annotated by the same ontology term. The C5 collection is divided into two subcollections, the first derived from the Gene Ontology resource (GO) which contains BP, CC, and MF components and a second derived from the Human Phenotype Ontology (HPO).

### Analysis of overexpressed genes
```{r define_upreg}
geneSets <- msigdbr(species = "Homo sapiens", category = "C5")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)
```

\newpage
#### GO: biological process
```{r GOBP_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
plot_GO(geneSetRef = geneSets, geneSet = overExpressed, category = "GOBP", showCategory = 20)
```

\newpage
#### GO: molecular function
```{r GOMF_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO molecular function
plot_GO(geneSetRef = geneSets, geneSet = overExpressed, category = "GOMF", showCategory = 20)
```

\newpage
#### GO: cellular compartment
```{r GOCC_up, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO cellular compartments
plot_GO(geneSetRef = geneSets, geneSet = overExpressed, category = "GOCC", showCategory = 20)
```

\newpage
### Analysis of underexpressed genes

#### GO: biological process
```{r GOBP_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
plot_GO(geneSetRef = geneSets, geneSet = underExpressed, category = "GOBP", showCategory = 20)
```

\newpage
#### GO: molecular function
```{r GOMF_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO molecular function
plot_GO(geneSetRef = geneSets, geneSet = underExpressed, category = "GOMF", showCategory = 20)
```

\newpage
#### GO: cellular compartment
```{r GOCC_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO cellular compartments
plot_GO(geneSetRef = geneSets, geneSet = underExpressed, category = "GOCC", showCategory = 20)
```

\newpage
## Regulatory elements (transcription factor binding and miRNAs)

### Regulatory: upregulated genes
```{r reg_up, dpi = 300, fig.width=13, fig.height= 11}
geneSets <- msigdbr(species = "Homo sapiens", category = "C3")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)

## Enrichment GO biological process
upReg_reg <- enricher(gene = overExpressed, TERM2GENE = geneSets)
dotplot(upReg_reg, showCategory = 20)
```

\newpage
### Regulatory: downregulated genes
```{r reg_down, dpi = 300, fig.width=13, fig.height= 11}
## Enrichment GO biological process
downReg_reg <- enricher(gene = underExpressed, TERM2GENE = geneSets)
dotplot(downReg_reg, showCategory = 20)
```









\newpage
# Gene set enrichment analysis (example)

This analysis is just an example to show the possibility of GSEA and depcition of GSEA plots if preferred to overrepresentation analysis.

```{r gsea}
## Deal with inf
allResGSEA <- allRes %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange))

## Remove NAs and order by GSEA
allResGSEA <- allResGSEA %>%
  filter(! is.na(gsea_metric)) %>%
  arrange(desc(gsea_metric))

## Get the ranked GSEA vector
ranks <- allResGSEA %>%
  select(SYMBOL, gsea_metric) %>%
  distinct(SYMBOL, .keep_all = TRUE) %>%
  deframe()

## define Gene sets (GO)
geneSets <- msigdbr(species = "Homo sapiens", category = "C5")
geneSets <- geneSets %>%
  select(gs_name, gene_symbol)

## Run GSEA
gseaRes <- GSEA(geneList = ranks, TERM2GENE = geneSets)
gseaResDf <- as.data.frame(gseaRes)
```

<!-- #### GO: biological process -->
<!-- ```{r GOBP_down, dpi = 300, fig.width=13, fig.height= 11} -->
<!-- ## Enrichment GO biological process -->
<!-- geneSets_GOBP <- geneSets[str_detect(geneSets$gs_name, "GOBP"),] -->
<!-- downReg_GOBP <- enricher(gene = underExpressed, TERM2GENE = geneSets_GOBP) -->
<!-- dotplot(downReg_GOBP, showCategory = 20) -->
<!-- ``` -->

```{r gsea_dotplot, dpi = 300, fig.width=10, fig.height= 12}
## Make summary plots
dotplot(gseaRes, showCategory = 20)
```

```{r gsea_testplot, dpi = 300, fig.width=7, fig.height= 5}
gseaplot(gseaRes, geneSetID = "GOCC_PIGMENT_GRANULE",
         title = "GOCC_PIGMENT_GRANULE")
```

\newpage
# Appenix

## Session Info
```{r}
utils::sessionInfo()
```
